---
title: "Plampa_john_rnaseq"
author: "Benjamin Young"
date: "14/12/2021"
output: html_document
---

OA experiment 
- contempory ph (8.05) versus 7.75
- AM = ambient
- OA = OA
- Colony is genotype (colony because have not genotyped)
- Tank, 2 tanks per treatment

1. PCA
2. DeSeq2 -> heatmaps
3. KEGG and GO enrichment 
4. Try WGCNA (it wont work)

When has other sponge -> eggnogmapper, gene name orthologs so we can compare the overlapping genes between the species and look at differences/similarities between the species. 


```{r, include=F}
library(tidyverse)
library(DESeq2)
library(edgeR)
library(limma)
library(Biostrings)
library(seqRFLP)
library(venn)
library(PCAtools)
library(DEGreport)
library(circlize)
library(ComplexHeatmap)
```


### Reading in Data

```{r , include = F}
read.csv(file = "~/Desktop/Experiment/PhD/Ch2_SpongeOA/DEG/DEG_Files/plampa/allcounts_host.csv", 
                   row.names = 1) -> counts

read.csv(file = "~/Desktop/Experiment/PhD/Ch2_SpongeOA/DEG/DEG_Files/plampa/metadata.csv", 
         sep = ",", 
         header = T) %>%
  mutate(sample_id = str_replace(sample_id, "P", "Pione_")) %>%
  column_to_rownames(var = "sample_id") -> tfall

tfall %>% 
  mutate(COLONY = as.factor(colony)) -> tfall

# View(tfall)
# View(counts)

nrow(tfall)
ncol(counts)
```


```{r , include = F}
tfall %>% 
  rownames_to_column(var = "sample_id") %>% 
#  dplyr::filter(!sample_id %in% c("Pione_9")) %>% 
  mutate(tank_fac = as.factor(tank)) %>%
  column_to_rownames(var = "sample_id") -> tfall
```


### Making Sexy Annotation File

```{r , include = F}
read.table(file = "~/Dropbox/PhD/people/john/plampa/Plampa-denovo-transcriptome-main/Plampa_iso2geneName.tab",
          sep = "\t") %>%
  mutate(GeneID = V1,
         Gene.Annotation = V2) %>%
  dplyr::select(-V1, -V2) %>%
  full_join(counts %>%
              rownames_to_column(var = "GeneID") %>%
              dplyr::select(GeneID)) %>%
  full_join(read.table("~/Dropbox/PhD/people/john/plampa/Plampa-denovo-transcriptome-main/Plampa_iso2go.tab",
            sep = "\t") %>%
              mutate(GeneID = V1,
                     GOID = V2) %>%
              dplyr::select(-V1, -V2)) %>%
  full_join(read.table("~/Dropbox/PhD/people/john/plampa/Plampa-denovo-transcriptome-main/Plampa_iso2kegg.tab",
            sep = "\t") %>%
              mutate(GeneID = V1,
                     KEGGID = V2) %>%
              dplyr::select(-V1, -V2)) %>%
  full_join(read.table("~/Dropbox/PhD/people/john/plampa/Plampa-denovo-transcriptome-main/Plampa_iso2kogClass.tab",
            sep = "\t") %>%
              mutate(GeneID = V1,
                     KOG.annotation = V2) %>%
              dplyr::select(-V1, -V2)) %>%
  full_join(read.table("~/Dropbox/PhD/people/john/plampa/Plampa-denovo-transcriptome-main/Plampa_seq2iso.tab",
            sep = "\t") %>%
              mutate(FastaID = V1,
                     GeneID = V2) %>%
              dplyr::select(-V1, -V2)) %>%
  mutate(across(everything(), ~ replace_na(.x, "-"))) -> plampa_annot

write.csv(plampa_annot,
          file = "~/Dropbox/PhD/people/john/plampa/plampa_annot_r.csv")
```

### Fasta File of just reads in count matrix 

```{r reading in fasta and organising, include = F}
readDNAStringSet(filepath = "~/Dropbox/PhD/people/john/plampa/Plampa (1).fasta") %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_header") %>%
  dplyr::rename(., sequence = x) %>%
  separate(gene_header, sep = " ", into = c("genename", "length", "path")) %>%
  dplyr::select(genename, sequence, length) %>%
  mutate(FastaID = str_replace(genename, "TRINITY_DN", "Plampa")) %>%
  mutate(genelength = str_length(sequence)) %>% #double checking length is same as in fasta, it is
  inner_join(plampa_annot %>%
               dplyr::select(FastaID, GeneID)) %>%
  group_by(GeneID) %>%
  slice_max(., length) %>%
  distinct(GeneID, .keep_all = T) -> plampa_sequence_to_gene

write.csv(plampa_sequence_to_gene, 
          file = "~/Dropbox/PhD/people/john/plampa/plampa_sequence_to_gene.csv")

plampa_sequence_to_gene %>%
  dplyr::filter(duplicated(GeneID) | duplicated(GeneID, fromLast = T)) #checking to make sure no duplicates, there are none. 
```

```{r , include = F}
nrow(plampa_sequence_to_gene)
nrow(counts)
```

Counts and the annot file are now the same wooooooooooooooo
Writing a fasta file using the sequence and the Plampaxxxx for eggnogmapper

```{r writing fasta for eggnog of all genes, include = F}
dataframe = data.frame(plampa_sequence_to_gene$GeneID,
                       plampa_sequence_to_gene$sequence)

df.fasta <- dataframe2fas(dataframe,
                          file = "~/Dropbox/PhD/people/john/plampa/plampadf.fasta")
```

```{r reading in eggnog results, include = F}
# read.table(file = "~/Dropbox/PhD/people/john/eggnog/out.emapper.annotations_4r",
#            sep = "\t",
#            header = T) %>% View()
# 
# read.csv(file = "~/Dropbox/PhD/people/john/eggnog/out.emapper.annotations.csv") %>% View()
```


## Gene Expression Analysis
### DeSeq2 Prep

```{r Matching, include = F}
matchup <- match(rownames(tfall), colnames(counts))
counts  <- counts[,matchup ]
all(rownames(tfall) == colnames(counts))
```

Count columns match treatment rownames yay

```{r Making DeSeq object, include=FALSE}
#making the dds model to use in deseq2
ddsall = DESeqDataSetFromMatrix(countData =  counts, 
                                tfall, 
                                ~ COLONY + treatment)
```

```{r CPM filtering and DDS object creation, echo = T}
nrow(ddsall)
ncol(ddsall)

# cpm filtering step and seeing what original VS filtered gene number left is
cccall <- counts(ddsall)
keep <- rowSums(cpm(cccall)>=1) >= 8
cccall <- cccall[keep, ]

nrow(ddsall)
ncol(ddsall)
nrow(cccall)
ncol(cccall)

cccall %>%
  as.data.frame()  -> cccall
  
ddsall <- DESeqDataSetFromMatrix(countData = cccall, 
                                 colData = tfall, 
                                 design = ~ COLONY + treatment)
```

I have done some pretty leniant filtering because fuck me 88k genes !!! For reference the Acropora palmata genome has ~32k
Working with 30k which is more manageable, shall see what deseq results and low counts produces

```{r More streamlined fasta for eggnog to take less time, include = F}
# View(plampa_sequence_to_gene)

plampa_sequence_to_gene %>% 
  dplyr::filter(GeneID %in% rownames(cccall)) -> filt_fasta_4_egg
  
dataframe = data.frame(filt_fasta_4_egg$GeneID,
                       filt_fasta_4_egg$sequence)

df.fasta <- dataframe2fas(dataframe,
                          file = "~/Dropbox/PhD/people/john/plampa/filt_fasta_4_egg.fasta")  
```

```{r readingi n streamlined eggnog fasta, include = F}

```


### PCA Analysis

```{r VST, include=FALSE}
## VSD stabilisation and visulisation
vsdall <- vst(ddsall, blind=FALSE)
```

```{r CLR transformation for PCAtools, include = F}
PCA_tools_all <- assay(vsdall)

rv <- rowVars(PCA_tools_all)
select <- order(rv, decreasing = TRUE)[seq_len(min(500, length(rv)))]

allsamps <- pca(PCA_tools_all[select,], metadata = tfall, removeVar = 0.10)
```

```{r Plots from PCAtools, fig.width=12, fig.height=6, include = F}
## Scree plot showing amount of variance explained by each PC (bars) and cumulative variance as you progress along bars (line)
screeplot(allsamps, 
          getComponents(allsamps, 1:15), 
          axisLabSize = 10, 
          titleLabSize = 10, 
          returnPlot = T, 
          ylim = c(0,90), 
          vline = c(findElbowPoint(allsamps$variance))) +
  geom_label(aes(x = findElbowPoint(allsamps$variance) + 1, y = 25,
      label = 'Elbow method', vjust = -1, size = 4))

# Shows the main ASV weights driving the variance
biplot(
  allsamps,
  colby = "treatment",
  hline = 0,
  vline = 0,
  legendPosition = 'right', 
  labSize = 2,
  pointSize = 2,
  legendLabSize = 6,
  legendTitleSize = 10,
  axisLabSize = 10
)
## Plotting of a large number of PCs
#pairsplot(allsamps)

## Plot showing the samples with the strongest loadings of ASVs for each axes identyfying ones which are driving the variance
#plotloadings(allsamps, labSize = 3)
```

```{r EigenPlots, fig.width=13, fig.height = 7}
# Plotting of the metadata to the PC axes to see which one has strong significant relationships with axes. 
eigencorplot(
  allsamps,
  components = getComponents(allsamps, 1:10),
  metavars = c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"),
  col = c('darkblue', 'blue2', 'black', 'red2', 'darkred'),
  cexCorval = 0.7,
  colCorval = 'white',
  fontCorval = 2,
  posLab = 'bottomleft',
  rotLabX = 45,
  posColKey = 'top',
  cexLabColKey = 1.5,
  scale = TRUE,
  main = 'PC1 - 11, Metadata Correlations',
  colFrame = 'white',
  plotRsquared = FALSE
)

eigencorplot(
  allsamps,
  components = getComponents(allsamps, 1:11),
  metavars = c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"),
  col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
  cexCorval = 1.2,
  fontCorval = 2,
  posLab = 'all',
  rotLabX = 45,
  scale = TRUE,
  main = bquote(
    Principal ~ Component ~ Pearson ~ r ^ 2 ~ metadata ~ significant ~ correlation
  ),
  plotRsquared = T,
  corFUN = 'pearson',
  corUSE = 'pairwise.complete.obs',
  corMultipleTestCorrection = 'BH',
  signifSymbols = c('****', '***', '**', '*', ''),
  signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1)
)
```

```{r, Correlation heatmap for dis, echo = F, fig.width=13, fig.height=4}
eigencorplot(
  allsamps,
  components = getComponents(allsamps, 1:11),
  metavars = c("treatment", "COLONY"),
  col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
  cexCorval = 2,
  fontCorval = 2,
  posLab = 'all',
  rotLabX = 45,
  scale = TRUE,
  main = bquote(
    Principal ~ Component ~ Pearson ~ r ^ 2 ~ metadata ~ significant ~ correlation
  ),
  plotRsquared = T,
  corFUN = 'pearson',
  corUSE = 'pairwise.complete.obs',
  corMultipleTestCorrection = 'BH',
  signifSymbols = c('****', '***', '**', '*', ''),
  signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1)
)
```


```{r PCA of the Samples, include = T}
## not transposing the CLR for the PCA
pca_samp <- prcomp(t(PCA_tools_all)[,select])
sample_loadings <- as.data.frame(pca_samp$x)
#View(sample_loadings)

## Summary of the PCA analysis
#biplot(pca_samp, cex=c(0.6,0.4), var.axes=F,scale=1, xlab=xlab, ylab=ylab)
summary(pca_samp)
#fviz_eig(pca_samp, addlabels = TRUE, ylim = c(0, 100), )
```

```{r Function for PC 2 and 3, include=FALSE}
#PCA 2 and 3 axis creation
pcaaxes23 = function (object, intgroup = "condition", ntop = 500, returnData = FALSE) 
{
    rv <- rowVars(assay(object))
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
        length(rv)))]
    pca <- prcomp(t(assay(object)[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(object)))) {
        stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <- as.data.frame(colData(object)[, intgroup, 
        drop = FALSE])
    group <- if (length(intgroup) > 1) {
        factor(apply(intgroup.df, 1, paste, collapse = ":"))
    }
    else {
        colData(object)[[intgroup]]
    }
    d <- data.frame(PC2 = pca$x[, 2], PC3 = pca$x[, 3], group = group, 
        intgroup.df, name = colnames(object))
    if (returnData) {
        attr(d, "percentVar") <- percentVar[2:3]
        return(d)
    }
    ggplot(data = d, aes_string(x = "PC2", y = "PC3", color = "group")) + 
        geom_point(size = 3) + xlab(paste0("PC2: ", round(percentVar[2] * 
        100), "% variance")) + ylab(paste0("PC3: ", round(percentVar[3] * 
        100), "% variance")) + coord_fixed()
}
```

```{r Function for PC 3 and 4, include=FALSE}
#PCA 3 and 4 axis creation
pcaaxes34 = function (object, intgroup = "condition", ntop = 500, returnData = FALSE) 
{
    rv <- rowVars(assay(object))
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
        length(rv)))]
    pca <- prcomp(t(assay(object)[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(object)))) {
        stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <- as.data.frame(colData(object)[, intgroup, 
        drop = FALSE])
    group <- if (length(intgroup) > 1) {
        factor(apply(intgroup.df, 1, paste, collapse = ":"))
    }
    else {
        colData(object)[[intgroup]]
    }
    d <- data.frame(PC3 = pca$x[, 3], PC4 = pca$x[, 4], group = group, 
        intgroup.df, name = colnames(object))
    if (returnData) {
        attr(d, "percentVar") <- percentVar[3:4]
        return(d)
    }
    ggplot(data = d, aes_string(x = "PC3", y = "PC4", color = "group")) + 
        geom_point(size = 3) + xlab(paste0("PC3: ", round(percentVar[3] * 
        100), "% variance")) + ylab(paste0("PC4: ", round(percentVar[4] * 
        100), "% variance")) + coord_fixed()
}
```

```{r Function for PC 4 and 5, include=FALSE}
#PCA 3 and 4 axis creation
pcaaxes45 = function (object, intgroup = "condition", ntop = 500, returnData = FALSE) 
{
    rv <- rowVars(assay(object))
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
        length(rv)))]
    pca <- prcomp(t(assay(object)[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(object)))) {
        stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <- as.data.frame(colData(object)[, intgroup, 
        drop = FALSE])
    group <- if (length(intgroup) > 1) {
        factor(apply(intgroup.df, 1, paste, collapse = ":"))
    }
    else {
        colData(object)[[intgroup]]
    }
    d <- data.frame(PC4 = pca$x[, 4], PC5 = pca$x[, 5], group = group, 
        intgroup.df, name = colnames(object))
    if (returnData) {
        attr(d, "percentVar") <- percentVar[4:5]
        return(d)
    }
    ggplot(data = d, aes_string(x = "PC4", y = "PC5", color = "group")) + 
        geom_point(size = 3) + xlab(paste0("PC4: ", round(percentVar[4] * 
        100), "% variance")) + ylab(paste0("PC5: ", round(percentVar[5] * 
        100), "% variance")) + coord_fixed()
}
```

```{r DeSeq2 PCAs for % loadings, include = F}
plotPCA(vsdall, intgroup=c("treatment"), returnData = F)
pcaaxes23(vsdall, intgroup=c("treatment"), returnData = F)
pcaaxes34(vsdall, intgroup=c("treatment"), returnData = F)
pcaaxes45(vsdall, intgroup=c("treatment"), returnData = F)
```

```{r DeSeq2 PCAs for % loadings, include = F}
plotPCA(vsdall, intgroup=c("colony"), returnData = F)
pcaaxes23(vsdall, intgroup=c("colony"), returnData = F)
pcaaxes34(vsdall, intgroup=c("colony"), returnData = F)
pcaaxes45(vsdall, intgroup=c("colony"), returnData = F)
```

```{r DeSeq2 PCAs for % loadings, include = F}
plotPCA(vsdall, intgroup=c("tank"), returnData = F)
pcaaxes23(vsdall, intgroup=c("tank"), returnData = F)
pcaaxes34(vsdall, intgroup=c("tank"), returnData = F)
pcaaxes45(vsdall, intgroup=c("tank"), returnData = F)
```

```{r DeSeq2 PCAs for % loadings, include = T}
plotPCA(vsdall, intgroup=c("bw"), returnData = F)
pcaaxes23(vsdall, intgroup=c("bw"), returnData = F)
pcaaxes34(vsdall, intgroup=c("bw"), returnData = F)
pcaaxes45(vsdall, intgroup=c("bw"), returnData = F)
```

```{r DeSeq2 PCAs for % loadings, include = F}
plotPCA(vsdall, intgroup=c("chemD"), returnData = F)
pcaaxes23(vsdall, intgroup=c("chemD"), returnData = F)
pcaaxes34(vsdall, intgroup=c("chemD"), returnData = F)
pcaaxes45(vsdall, intgroup=c("chemD"), returnData = F)
```

```{r DeSeq2 PCAs for % loadings, include = F}
plotPCA(vsdall, intgroup=c("chemN"), returnData = F)
pcaaxes23(vsdall, intgroup=c("chemN"), returnData = F)
pcaaxes34(vsdall, intgroup=c("chemN"), returnData = F)
pcaaxes45(vsdall, intgroup=c("chemN"), returnData = F)
```

```{r DeSeq2 PCAs for % loadings, include = F}
plotPCA(vsdall, intgroup=c("mech"), returnData = F)
pcaaxes23(vsdall, intgroup=c("mech"), returnData = F)
pcaaxes34(vsdall, intgroup=c("mech"), returnData = F)
pcaaxes45(vsdall, intgroup=c("mech"), returnData = F)
```

```{r PCA Objects to use in GGPlot2, include = F}
pca12 <- plotPCA(vsdall, intgroup=c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"), returnData = TRUE)
pca23 <- pcaaxes23(vsdall, intgroup=c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"), returnData = TRUE)
pca34 <- pcaaxes34(vsdall, intgroup=c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"), returnData = TRUE)
pca45 <- pcaaxes45(vsdall, intgroup=c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"), returnData = TRUE)
```

```{r}
ggplot(pca12, aes(PC1, PC2, color = COLONY, shape = treatment)) +
  geom_point(size = 3) +  
  xlab(paste0("PC1 39% variance")) +
  ylab(paste0("PC2 23% variance")) +
  theme(
    text = element_text(size = 11, family = "Arial"),
    legend.position = "right",
    panel.background = element_rect(fill = "transparent"),
    axis.line = element_line(colour = "black"),
    legend.key = element_rect(fill = "transparent"),
    axis.text = element_text(size = 8)
  )  +
  theme(legend.key.size = unit(0.7, "cm")) +
#  geom_polygon(fill = "NA") +
  scale_color_manual(values = c("gold3", "orchid3", "aquamarine3")) + 
  stat_ellipse(aes(PC1, PC2, group = COLONY), type = "norm")
  

ggplot(pca23, aes(PC2, PC3, color = COLONY, shape = treatment)) + 
  geom_point(size=3) +  
  xlab(paste0("PC2 23% variance")) + 
  ylab(paste0("PC3 11% variance")) + 
  theme(text = element_text(size=11, family = "Arial"), 
        legend.position = "right", 
        panel.background = element_rect(fill = "transparent"), 
        axis.line = element_line(colour = "black"), 
        legend.key = element_rect(fill = "transparent"), 
        axis.text = element_text(size = 8)) + 
  geom_polygon(fill = "NA")

ggplot(pca34, aes(PC3, PC4, color = COLONY, shape = treatment)) + 
  geom_point(size=3) +  
  xlab(paste0("PC3 11% variance")) + 
  ylab(paste0("PC4 6% variance")) + 
  theme(text = element_text(size=11, family = "Arial"), 
        legend.position = "right", 
        panel.background = element_rect(fill = "transparent"), 
        axis.line = element_line(colour = "black"), 
        legend.key = element_rect(fill = "transparent"), 
        axis.text = element_text(size = 8)) + 
  geom_polygon(fill = "NA")

ggplot(pca45, aes(PC4, PC5, color = COLONY, shape = treatment)) + 
  geom_point(size=3) +  
  xlab(paste0("PC4 6% variance")) + 
  ylab(paste0("PC5 4% variance")) + 
  theme(text = element_text(size=11, family = "Arial"), 
        legend.position = "right", 
        panel.background = element_rect(fill = "transparent"), 
        axis.line = element_line(colour = "black"), 
        legend.key = element_rect(fill = "transparent"), 
        axis.text = element_text(size = 8)) + 
  geom_polygon(fill = "NA")
```


### PCA removing Colony Variance

```{r}
vsdgenet <- vst(ddsall)
assay(vsdgenet) <- limma::removeBatchEffect(assay(vsdgenet), vsdgenet$COLONY)
```

```{r CLR transformation for PCAtools, include = F}
PCA_tools_all <- assay(vsdgenet)

rv <- rowVars(PCA_tools_all)
select <- order(rv, decreasing = TRUE)[seq_len(min(500, length(rv)))]

allsamps <- pca(PCA_tools_all[select,], metadata = tfall, removeVar = 0.10)
```

```{r Plots from PCAtools, fig.width=12, fig.height=6, include = F}
## Scree plot showing amount of variance explained by each PC (bars) and cumulative variance as you progress along bars (line)
screeplot(allsamps, 
          getComponents(allsamps, 1:15), 
          axisLabSize = 10, 
          titleLabSize = 10, 
          returnPlot = T, 
          ylim = c(0,90), 
          vline = c(findElbowPoint(allsamps$variance))) +
  geom_label(aes(x = findElbowPoint(allsamps$variance) + 1, y = 25,
      label = 'Elbow method', vjust = -1, size = 4))

# Shows the main ASV weights driving the variance
biplot(
  allsamps,
  colby = "treatment",
  hline = 0,
  vline = 0,
  legendPosition = 'right', 
  labSize = 2,
  pointSize = 2,
  legendLabSize = 6,
  legendTitleSize = 10,
  axisLabSize = 10
)
## Plotting of a large number of PCs
#pairsplot(allsamps)

## Plot showing the samples with the strongest loadings of ASVs for each axes identyfying ones which are driving the variance
#plotloadings(allsamps, labSize = 3)
```

```{r EigenPlots, fig.width=13, fig.height = 7}
# Plotting of the metadata to the PC axes to see which one has strong significant relationships with axes. 
eigencorplot(
  allsamps,
  components = getComponents(allsamps, 1:10),
  metavars = c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"),
  col = c('darkblue', 'blue2', 'black', 'red2', 'darkred'),
  cexCorval = 0.7,
  colCorval = 'white',
  fontCorval = 2,
  posLab = 'bottomleft',
  rotLabX = 45,
  posColKey = 'top',
  cexLabColKey = 1.5,
  scale = TRUE,
  main = 'PC1 - 11, Metadata Correlations',
  colFrame = 'white',
  plotRsquared = FALSE
)

eigencorplot(
  allsamps,
  components = getComponents(allsamps, 1:11),
  metavars = c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"),
  col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
  cexCorval = 1.2,
  fontCorval = 2,
  posLab = 'all',
  rotLabX = 45,
  scale = TRUE,
  main = bquote(
    Principal ~ Component ~ Pearson ~ r ^ 2 ~ metadata ~ significant ~ correlation
  ),
  plotRsquared = T,
  corFUN = 'pearson',
  corUSE = 'pairwise.complete.obs',
  corMultipleTestCorrection = 'BH',
  signifSymbols = c('****', '***', '**', '*', ''),
  signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1)
)
```

```{r DeSeq2 PCAs for % loadings}
plotPCA(vsdgenet, intgroup=c("treatment"), returnData = F)
pcaaxes23(vsdgenet, intgroup=c("treatment"), returnData = F)
pcaaxes34(vsdgenet, intgroup=c("treatment"), returnData = F)
pcaaxes45(vsdgenet, intgroup=c("treatment"), returnData = F)
```

```{r DeSeq2 PCAs for % loadings, include = F}
plotPCA(vsdgenet, intgroup=c("tank"), returnData = F)
pcaaxes23(vsdgenet, intgroup=c("tank"), returnData = F)
pcaaxes34(vsdgenet, intgroup=c("tank"), returnData = F)
pcaaxes45(vsdgenet, intgroup=c("tank"), returnData = F)
```

```{r PCA Objects to use in GGPlot2, include = F}
pca12 <- plotPCA(vsdgenet, intgroup=c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"), returnData = TRUE)
pca23 <- pcaaxes23(vsdgenet, intgroup=c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"), returnData = TRUE)
pca34 <- pcaaxes34(vsdgenet, intgroup=c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"), returnData = TRUE)
pca45 <- pcaaxes45(vsdgenet, intgroup=c("treatment", "COLONY", "tank", "bw", "chemD", "chemN", 
                                       "mech"), returnData = TRUE)
```

```{r PCA plots of treatment, include = F}
ggplot(pca12, aes(PC1, PC2, color = treatment)) +
  geom_point(size = 3) +  
  xlab(paste0("PC1 27% variance")) +
  ylab(paste0("PC2 20% variance")) +
  theme(
    text = element_text(size = 11, family = "Arial"),
    legend.position = "right",
    panel.background = element_rect(fill = "transparent"),
    axis.line = element_line(colour = "black"),
    legend.key = element_rect(fill = "transparent"),
    axis.text = element_text(size = 8)
  )  +
  theme(legend.key.size = unit(0.7, "cm")) + 
  stat_ellipse()

ggplot(pca23, aes(PC2, PC3, color = treatment)) + 
  geom_point(size=3) +  
  xlab(paste0("PC2 20% variance")) + 
  ylab(paste0("PC3 10% variance")) + 
  theme(text = element_text(size=11, family = "Arial"), 
        legend.position = "right", 
        panel.background = element_rect(fill = "transparent"), 
        axis.line = element_line(colour = "black"), 
        legend.key = element_rect(fill = "transparent"), 
        axis.text = element_text(size = 8)) + 
  stat_ellipse()

ggplot(pca34, aes(PC3, PC4, color = treatment)) + 
  geom_point(size=3) +  
  xlab(paste0("PC3 10% variance")) + 
  ylab(paste0("PC4 7% variance")) + 
  theme(text = element_text(size=11, family = "Arial"), 
        legend.position = "right", 
        panel.background = element_rect(fill = "transparent"), 
        axis.line = element_line(colour = "black"), 
        legend.key = element_rect(fill = "transparent"), 
        axis.text = element_text(size = 8)) + 
  stat_ellipse()

ggplot(pca45, aes(PC4, PC5, color = treatment)) + 
  geom_point(size=3) +  
  xlab(paste0("PC4 7% variance")) + 
  ylab(paste0("PC5 6% variance")) + 
  theme(text = element_text(size=11, family = "Arial"), 
        legend.position = "right", 
        panel.background = element_rect(fill = "transparent"), 
        axis.line = element_line(colour = "black"), 
        legend.key = element_rect(fill = "transparent"), 
        axis.text = element_text(size = 8)) + 
  stat_ellipse()
```

### Running DeSeq

I am a little unsure what to look at, but I think there are some routes to go 

1. Looking at the differences between the 3 genets/colonies. See what driving colony identity 
 - model probably using the LRT methodology to get all significant genes between the colonies
 - can then look at specific difs between genets 
2. Seeing if we can get some difs between OA and AMB 
 - model would be ~ colony + treatment so to control for the colony identity and get the major gene shifts due to treatment 
 
#### Genet Identity 

```{r Differential Exprssion analysis, fig.width=9, fig.height=9}
ddslrt <- DESeqDataSetFromMatrix(countData = cccall, 
                                 colData = tfall, 
                                 design = ~ COLONY)

ddslrt <- DESeq(ddslrt, test="LRT", reduced = ~ 1)
lrt_results <- results(ddslrt, 
                       alpha = 0.001)

summary(lrt_results)
```

```{r}
# as.data.frame(lrt_results) %>% 
#   na.exclude() %>% 
#   dplyr::filter(padj <= 0.001) %>%
#   rownames_to_column(var = "GeneID") %>% 
#   inner_join(plampa_annot) -> LRT_geno_0.001
# 
# LRT_geno_0.001

# write.csv(LRT_geno_0.001,
#           file = "~/Dropbox/PhD/Projects/DHE/NGS/tagseq/dissertation_files/raw_lists/LRT_geno_0.001.csv")

# LRT_geno_0.01 %>% 
#   rownames() -> filt_lrt
```

```{r contrasts between each colony}
resultsNames(ddslrt)

C_A_res <- results(ddslrt, 
                       alpha = 0.001,
                       test = "Wald", 
                       contrast = c("COLONY", "C", "A"))

C_B_res <- results(ddslrt, 
                       alpha = 0.001,
                       test = "Wald", 
                       contrast = c("COLONY", "C", "B"))

A_B_res <- results(ddslrt, 
                       alpha = 0.001,
                       test = "Wald", 
                       contrast = c("COLONY", "A", "B"))
```

```{r}
summary(C_A_res)
as.data.frame(C_A_res) %>%
  na.exclude() %>%
  filter(padj < 0.001) %>%
  mutate(DeSeq_contrast = "c_vs_a") -> C_A_res_0.01

summary(C_B_res)
as.data.frame(C_B_res) %>%
  na.exclude() %>%
  filter(padj < 0.001) %>%
  mutate(DeSeq_contrast = "c_vs_b") -> C_B_res_0.01

summary(A_B_res)
as.data.frame(A_B_res) %>%
  na.exclude() %>%
  filter(padj < 0.001) %>%
  mutate(DeSeq_contrast = "a_vs_b") -> A_B_res_0.01
```


```{r}
venn(list(rownames(C_A_res_0.01), rownames(A_B_res_0.01), rownames(C_B_res_0.01)))
```

Genet identity has a strong signal (As expected) 

At alpha 0.001, decent amount of significant genes (See output from above). Around 11,884 genes.  
  
  
##### Clusterprofiler Analysis

This analysis identifies similar expression profiles of genes from the LRT (N.B. LRT is all significant genes between the 3 genets here, like a  one-way anova) and places them into clusters.  
  
We can therfore see similar expression profiles for sets of genes and if they are higher/lower in different genets.  

```{r Differential Exprssion analysis, fig.width=9, fig.height=9}
summary(lrt_results)

as.data.frame(lrt_results) %>%
  na.exclude() %>%
  dplyr::filter(padj < 0.0001) -> LRT_0.01

LRT_0.01 %>% 
  rownames() -> filt_lrt
# View(filt_lrt)
```

```{r saving DEG results, fig.height=5, fig.width=7}
# assay(vsdall) %>%
#  as.data.frame() %>%
#  rownames_to_column(var = "genes") %>%
#  dplyr::filter(genes %in% filt_lrt) %>%
#  column_to_rownames(var = "genes") -> ma
# 
# design <- as.data.frame(colData(ddslrt))
# 
# res_norm <- degPatterns(ma, tfall, time = "COLONY", reduce = T)
# res_norm_scale <- degPatterns(ma, tfall, time = "COLONY", reduce = T, scale = T)
# 
# save(res_norm,
#     file = "~/Dropbox/PhD/people/john/plampa/DEG_Report_redT.RData")
# save(res_norm_scale,
#     file = "~/Dropbox/PhD/people/john/plampa//DEG_Report_redT_scaleT.RData")

load("~/Dropbox/PhD/people/john/plampa/DEG_Report_redT.RData")
load("~/Dropbox/PhD/people/john/plampa//DEG_Report_redT_scaleT.RData")

degPlotCluster(res_norm$normalized, time = "colony", points = F) + theme_bw()
degPlotCluster(res_norm_scale$normalized, time = "colony", points = F) + theme_bw()
```

```{r LRT Cluster ggplot, echo = T}
ggplot(res_norm$normalized, aes(x=COLONY, y=value, color = COLONY)) + 
  geom_boxplot(outlier.size = 0, outlier.shape = NA) + 
  stat_smooth(aes(x = COLONY, y = value, 
            group = COLONY, color = COLONY), se = FALSE, method = "lm", 
            formula = y ~ poly(x, splan)) +
  geom_line(data = res_norm$normalized %>% 
                group_by(COLONY, cluster) %>% 
                summarise(average = mean(value)) %>% 
                ungroup(),
            mapping = aes(x=COLONY, y = average, colour = COLONY), 
            color = "black") + 
  geom_line(data = res_norm$normalized %>% 
                group_by(COLONY, cluster) %>% 
                summarise(average = mean(value)) %>% 
                ungroup(),
            mapping = aes(x=COLONY, y = average, group = 0)) +
  facet_wrap(~cluster) +
  theme_bw()
```

```{r writing clusters to csvs, include = F}
res_norm_scale$df %>%
  filter(cluster %in% c("1")) %>%
  write.csv(., file = "/Users/benyoung/Dropbox/PhD/people/john/plampa/clupro_results/LRT_clus1.csv")

res_norm_scale$df %>%
  filter(cluster %in% c("2")) %>%
  write.csv(., file = "/Users/benyoung/Dropbox/PhD/people/john/plampa/clupro_results/LRT_clus2.csv")

res_norm_scale$df %>%
  filter(cluster %in% c("3")) %>%
  write.csv(., file = "/Users/benyoung/Dropbox/PhD/people/john/plampa/clupro_results/LRT_clus3.csv")

res_norm_scale$df %>%
  filter(cluster %in% c("4")) %>%
  write.csv(., file = "/Users/benyoung/Dropbox/PhD/people/john/plampa/clupro_results/LRT_clus4.csv")
```


  
#### OA Versus AMB

I do not think this will be worth including, its minimal and there is not much to look at tbh. There may be some genes but as you said from your physio data you get the asymptotic curve, so it kinda makses sense that it is similar to the ambient OA.  
  
I have included for wholeness.  
  
```{r Differential Exprssion analysis Disease Response, include = F}
dds_OA <- DESeqDataSetFromMatrix(countData = cccall, 
                                      colData = tfall, 
                                      design = ~ colony + treatment)

dds_OA <- DESeq(dds_OA, test="Wald")

resultsNames(dds_OA)

OA_versus_AMB <- results(dds_OA, 
                   contrast = c("treatment", "OA", "AM"), 
                   alpha = 0.05, 
                   test = "Wald")
```

```{r OA results, echo = T}
summary(OA_versus_AMB)
```

```{r, include = F}
as.data.frame(OA_versus_AMB) %>% 
  dplyr::filter(padj < 0.05) -> OA_versus_AMB_res_0.05
```

```{r fasta file for eggnog}
View(OA_versus_AMB_res_0.05)
View(plampa_sequence_to_gene)

plampa_sequence_to_gene %>% 
  dplyr::filter(GeneID %in% rownames(OA_versus_AMB_res_0.05)) %>% 
  dplyr::select(GeneID, sequence) -> df_4_blast_OA

df = data.frame(df_4_blast_OA$GeneID, 
                df_4_blast_OA$sequence)

df.fasta <- dataframe2fas(df,
                          file = "~/Dropbox/PhD/people/john/plampa/OA_signif.fasta")
```



##### Heatmap OA Genes

```{r}
plampa_annot %>% 
  dplyr::filter(FastaID %in% plampa_sequence_to_gene$FastaID) %>%
  dplyr::mutate(test = str_replace_all(Gene.Annotation, 
                                       pattern = "^-$", 
                                       replacement = "hypothetical_protein")) -> plampa_annot_GENES
```


```{r Signif Genes we want, include=FALSE}
forheatmap <- as.character(rownames(OA_versus_AMB_res_0.05))
```

```{r All common 68 genes prep, include = F}
matmatnorder <- assay(vsdgenet)[forheatmap,]

# Colours for the tank treatments
ccann <- data.frame(tfall$treatment)
colnames(ccann) <- c("treatment")
colcol <- list("treatment" = c("AM" = "dodgerblue", "OA" = "red4"))

# Top annotation bar for the complex heatmap
samname <- HeatmapAnnotation(df = ccann,
                             which = "col",
                             col = colcol,                        
                             simple_anno_size = unit(0.5, "cm"),
                             annotation_name_gp = gpar(fontsize = 10),
                             annotation_name_side = "left")

# Fill for the gene VST data
col_fun = colorRamp2(c(-2, 0, 2), 
                     c("yellow3", "white", "darkorchid4"))
col_fun(seq(-3, 3))

# adding annotation into vsd count matrix
plampa_annot_GENES %>%
  dplyr::select(GeneID, Gene.Annotation) %>%
  right_join(matmatnorder %>%
               as.data.frame() %>%
               rownames_to_column(var="GeneID")) %>%
  column_to_rownames(var="GeneID") %>%
  tidyr::drop_na() -> matmatnorder
#View(matmatnorder)
# 
# # Changing names to a character
# matmatnorder$Gene.Annotation <- as.character(matmatnorder$Gene.Annotation)
# 
# #id_to_annot, need to have rownames when putting into complex heatmap
# hello <-
#   rowAnnotation(foo = anno_text(
#     matmatnorder %>% as.data.frame() %>% dplyr::select(Gene.Annotation)
#   ))
```

```{r PC1 Variance Heatmaps, fig.height=5, fig.width=5}
# NB manual specify of the GO columns, here it is 86:93. First heatmap we remove, 2nd heatmap we select. 
  Heatmap(
    t(scale(
      t(matmatnorder %>% as.data.frame() %>% dplyr::select(-Gene.Annotation))
    )),
    cluster_columns = T,
    show_row_dend = F,
    show_column_dend = T,
    column_km = 2,
    column_gap = unit(0.3, "cm"),
    row_km = 2,
    row_gap = unit(0.3, "cm"),
    col = col_fun,
    row_title = NULL,
    column_title = NULL,
    top_annotation = samname,
    show_row_names = F, 
    show_column_names = F,
    column_dend_height = unit(1.5, "cm"), 
    right_annotation = rowAnnotation(foo = anno_text(
      matmatnorder$Gene.Annotation,
      gp = gpar(fontsize = 6, font = 10))))
```
